core/interval/int.go

```diff
@@ -0,0 +1,117 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package interval
+
+import (
+	"sort"
+)
+
+type IntegerInterval struct {
+	Lower int
+	Upper int
+}
+
+func NewIntegerInterval(a, b int) IntegerInterval {
+	return IntegerInterval{
+		Lower: min(a, b),
+		Upper: max(a, b),
+	}
+}
+
+func (i IntegerInterval) Intersects(other IntegerInterval) bool {
+	return i.Lower <= other.Upper && i.Upper >= other.Lower
+}
+
+func (i IntegerInterval) Adjacent(other IntegerInterval) bool {
+	return i.Upper+1 == other.Lower || i.Lower-1 == other.Upper
+}
+
+func (i IntegerInterval) IsSubsetOf(other IntegerInterval) bool {
+	return i.Lower >= other.Lower && i.Upper <= other.Upper
+}
+
+func (i IntegerInterval) Difference(other IntegerInterval) IntegerIntervals {
+	if !i.Intersects(other) {
+		return []IntegerInterval{i}
+	}
+
+	if i.IsSubsetOf(other) {
+		return []IntegerInterval{}
+	}
+
+	// We shouldn't just use IsSupersetOf because it would return true for
+	// intervals which match at one end, meaning the difference only on
+	// one side. This is covered in later cases.
```

Comment:
This comment is a bit confusing as a reader, I think it requires a lot of implied context to understand.

Reply:
I have re-written